algorithms/linfa-clustering/src/optics/algorithm.rs

```diff
@@ -0,0 +1,291 @@
+use crate::optics::hyperparameters::{OpticsHyperParams, OpticsHyperParamsBuilder};
+use float_ord::FloatOrd;
+use hnsw::{Params, Searcher, HNSW};
+use linfa::traits::Transformer;
+use linfa::Float;
+use ndarray::{ArrayBase, ArrayView, ArrayView1, Axis, Data, Ix1, Ix2};
+use ndarray_stats::DeviationExt;
+#[cfg(feature = "serde")]
+use serde_crate::{Deserialize, Serialize};
+use space::MetricPoint;
+use std::cmp::{max, Ordering};
+use std::collections::BTreeSet;
+
+/// Implementation of euclidean distance for ndarray
+struct Euclidean<'a, F>(ArrayView1<'a, F>);
```

Comment:
there is also the plan to unify these in `linfa-nn` https://github.com/rust-ml/linfa/issues/119#issuecomment-841839071

Reply:
Once they're unified I can change over, otherwise if this is done and merged first I guess it will be unified when the great unification happens

Reply:
sure!